adding kube2iam experimental support

Fsero opened the pull request:
If you plan to use kube2iam in existing kubeaws cluster you face two things:

- Default WorkerRole does not have the sts:assumeRole grant blocking kube2iam to assume the role for the container.
- target roles needs to change the trust policy to allow kube-aws worker IAM Role to assume the role. Because normally cloudformation generates the role with the role name and a random ID at the end makes the IAM Role less predictable and therefore more complicated to work with if you want to automate this trust.

This pr introduces kube2iam experimental flag, when kube2iam flag is enabled it will add sts:assumeRole to the worker IAM role to allow kube2iam to work, additionally you can specified a named role to generate a worker IAM Role more human and automation friendly as explained here http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html#cfn-iam-role-rolename


If you find the named role useful maybe could be merged in the general configuration not only for the worker

Fsero commented:
i think this also helps issue #253

mumoshu commented on config/templates/cluster.yaml:
```diff
@@ -469,6 +469,12 @@ kmsKeyArn: "{{.KMSKeyARN}}"
+#    # When enabled it will grant sts:assumeRole permission to everyrole in the IAM Worker Role.
+#    # if you specify workerNamedIAMRole the role created will not suffix the random id at the end
+#    # which is useful to update trust policy in target IAM roles that container wants to use through kube2iam.
+#    kube2IAMSupport:
```
nit but it should be spelled `kube2IamSupport` for consistency among other keys in cluster.yaml

mumoshu commented:
Thanks for the PR with very informative description, @Fsero!
This is almost merged but could you add an example configuration to the test data for the case named `WithExperimentalFeatures` in `test/integration/maincluster_test.go` so that we can ensure that this feature won't break in the future?

Fsero commented:
@mumoshu thanks! added the tests in `maincluster_test.go` and `nodepool_test.go` and changed the case of every IAM mention. I hope this is enough, in other case let me know what can i do to get this merged :)

pieterlange commented on config/templates/cluster.yaml:
```diff
@@ -469,6 +469,12 @@ kmsKeyArn: "{{.KMSKeyARN}}"
+#    # which is useful to update trust policy in target IAM roles that container wants to use through kube2iam.
+#    kube2IamSupport:
+#      enabled: true
+#      workerNamedIAMRole: "yournamedrole"
```
What is the point of letting people choose the role name here? We don't let them choose resource names elsewhere :)

Fsero replied:
@pieterlange if you have a target role for kube2iam ,let say role A, if role A has a trust policy for role R set in cluster.yaml the cluster recreation does not need to retriever a trust policy update

Fsero replied:
S/retriever/retrigger/g

Fsero replied:
edited in web because it was not clear,  check it if you had read email.

c-knowles replied:
Could we just fix the name based on the stack name postfixed with some static role name? Perhaps even [export it](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-stack-exports.html) so it can be referenced from other stacks?

Fsero replied:
@c-knowles one thing i can do is to change the namedRole to something like {{stack-name}}-IAMRoleWorker-{{suffixRoleName}, and then change  workerNamedIAMRole to workerNamedIAMRoleSuffix and set a sensible default like kube2iam, i would prefer to give the users the option to set the naming it makes most sense in their scenario.

c-knowles replied:
It'll work either way so I don't mind, just the same as @pieterlange I'm wondering what the use case is for allowing it to be changed rather than just fixing it against the stack name.

pieterlange commented:
After this is merged we should also:
  * add some docs that show how to use kube2iam
  * remove all default controller/worker IAM permissions that are not strictly required (create them as roles and attach them to the kube pods that actually need them)

Great work!

c-knowles commented:
@pieterlange if we are going to use it for other purposes then perhaps when it's not experimental any more we should consider renaming it from `Kube2IamSupport` to something like `roleName` under the `worker` config?

Maybe we could also consider whether to allow a list of roles that are allowed to be assumed from a k8s cluster or just always allow `*`.

mumoshu commented:
Note to self: I'd rather like to move `experimental.kube2IamSupportworkerNamedIAMRole` to `worker.managedIamRoleName` so that we can let users generally use it to trust the role managed for worker nodes by kube-aws from a target role.

Fsero commented:
@mumoshu is there something else I can do to get this merged? :-)

mumoshu commented:
@Fsero Sorry for the silence!
I'm still busy working on my own PR #315 but in the meantime would you confirm/apply what I've suggested in https://github.com/coreos/kube-aws/pull/297#issuecomment-278492922 on your side? (Btw, does the suggested name `worker.managedIamRoleName` over the former `kube2IamSupportWorkerNamedIAMRole` make sense to you too?)

Anyways, this PR is going to be merged before my own PR not to delegate the hard work of resolving huge conflicts 👍

Fsero commented:
sure @mumoshu i can do that :)

mumoshu commented:
Thanks @Fsero!
Could you also rebase your branch? FYI/AFAICS `ClusterAutoscalerSupport ClusterAutoscalerSupport `yaml:"clusterAutoscalerSupport"` from an another PR is the only line conflicting with yours.

Fsero commented:
@mumoshu i rewrited my PR with the proposed changes, i kept the experimental feature for kube2iam because you maybe want to have the managedIamRole without using kube2iam, in doesnt make sense we can always drop this feature flag and add this grant in the base IAM Role

mumoshu commented:
LGTM 👍 Thanks for your contribution @Fsero!
